riscv/csrs.cc

```diff
@@ -1400,3 +1406,30 @@ void henvcfg_csr_t::verify_permissions(insn_t insn, bool write) const {
       throw trap_illegal_instruction(insn.bits());
   }
 }
+
+stimecmp_csr_t::stimecmp_csr_t(processor_t* const proc, const reg_t addr, const reg_t imask):
+  basic_csr_t(proc, addr, 0), intr_mask(imask) {
+}
+
+bool stimecmp_csr_t::unlogged_write(const reg_t val) noexcept {
+  state->mip->backdoor_write_with_mask(intr_mask, state->time->read() >= val ? intr_mask : 0);
+  return basic_csr_t::unlogged_write(val);
+}
+
+virtualized_stimecmp_csr_t::virtualized_stimecmp_csr_t(processor_t* const proc, csr_t_p orig, csr_t_p virt):
+  virtualized_csr_t(proc, orig, virt) {
+}
+
+void virtualized_stimecmp_csr_t::verify_permissions(insn_t insn, bool write) const {
+  // check for read permission to time as enabled by xcounteren
+  state->time_proxy->verify_permissions(insn, false);
+
```

Comment:
This seems not right. CSR_TIME is a user mode CSR. If only  verify_permissions for CSR_TIME is reused here, access from U/VU mode for stimecmp/vstimecmp may be allowed.

Reply:
But this method does not call its own base class's `verify_permission()`, only the `verify_permission()` of `time_proxy`, which will check the U-mode permissions appropriate to `time`. (Right?)

Reply:
`timer_proxy->verify_permission` -> `counter_proxy_csr_t::verify_permissions` -> `proxy_csr_t::verify_permissions` -> `csr_t::verify_permissions`, which should throw an exception if U mode is accessing S/HS mode CSRs.

`counter_proxy_csr_t::verify_permissions` then does extra checking of xcounteren and throws the appropriate exception.

Reply:
The `time_proxy` object has a CSR address of `CSR_TIME` and therefore a `csr_priv` of `PRV_U` so its parent classes' methods will never check the access permissions of `CSR_STIMECMP` or `CSR_VSTIMECMP`.

Reply:
You're right. I need to call `virtualized_csr_t::verify_permissions` explicitly.